File: src/CVNBot/RCReader.cs
using System;
using System.Text.RegularExpressions;
using System.Threading;
using log4net;
using Meebey.SmartIrc4net;

namespace CVNBot
{
    struct RCEvent
    {
        public enum EventType
        {
            delete, restore, upload, block, unblock, edit, protect, unprotect,
            move, rollback, newuser, import, unknown, newuser2, autocreate,
            modifyprotect
        }

        public string project;
        public string title;
        public string url;
        public string user;
        public bool minor;
        public bool newpage;
        public bool botflag;
        public int szdiff;
        public string comment;
        public EventType eventtype;
        public string blockLength;
        public string movedTo;

        public override string ToString()
        {
            return "[" + project + "] " + user + " edited [[" + title + "]] (" + szdiff.ToString() + ") " + url + " " + comment;
        }
    }

    class RCReader
    {
        public IrcClient rcirc = new IrcClient();
        public DateTime lastMessage = DateTime.Now;

        // RC parsing regexen
        static readonly Regex stripColours = new Regex(@"\x04\d{0,2}\*?");
        static readonly Regex stripColours2 = new Regex(@"\x03\d{0,2}");
        static readonly Regex stripBold = new Regex(@"\x02");
        static readonly Regex rszDiff = new Regex(@"\(([\+\-])([0-9]+)\)");

        static readonly ILog logger = LogManager.GetLogger("CVNBot.RCReader");

        static readonly string serverName = "irc.wikimedia.org";

        public void InitiateConnection()
        {
            Thread.CurrentThread.Name = "RCReader";

            logger.Info("Thread started");

            // Set up RCReader
            rcirc.Encoding = System.Text.Encoding.UTF8;
            rcirc.AutoReconnect = true;
            rcirc.AutoRejoin = true;

            rcirc.OnChannelMessage += Rcirc_OnChannelMessage;
            rcirc.OnConnected += Rcirc_OnConnected;

            try
            {
                rcirc.Connect(serverName, 6667);
            }
            catch (ConnectionException e)
            {
                logger.Warn("Could not connect", e);
                return;
            }

            try
            {
                rcirc.Login(Program.config.botNick, "CVNBot", 4, "CVNBot");

                logger.InfoFormat("Joining {0} channels", Program.prjlist.Count);
                foreach (string prj in Program.prjlist.Keys)
                {
                    rcirc.RfcJoin("#" + prj);
                }

                // Enter loop
                rcirc.Listen();
                // When Listen() returns the IRC session is over
                rcirc.Disconnect();
            }
            catch (ConnectionException)
            {
                // Final disconnect may throw, ignore.
                return;
            }
        }

        void Rcirc_OnConnected(object sender, EventArgs e)
        {
            logger.InfoFormat("Connected to {0}", serverName);
        }

        void Rcirc_OnChannelMessage(object sender, IrcEventArgs e)
        {
            lastMessage = DateTime.Now;

            // Based on RCParser.py->parseRCmsg()
            // Example message from 2017-10-13 from #en.wikipedia
            // 01> #00314 [[
            // 02> #00307 Special:Log/newusers
            // 03> #00314 ]]
            // 04> #0034   create2
            // 05> #00310
            // 06> #00302
            // 07> #003
            // 08> #0035  *
            // 09> #003
            // 10> #00303 Ujju.19788
            // 11> #003
            // 12> #0035  *
            // 13> #003
            // 14> #00310 created new account User:Upendhare
            // 15> #003
            string strippedmsg = stripBold.Replace(stripColours.Replace(CVNBotUtils.ReplaceStrMax(e.Data.Message, '\x03', '\x04', 14), "\x03"), "");
            string[] fields = strippedmsg.Split(new char[] { '\x03' }, 15);
            if (fields.Length == 15)
            {
                if (fields[14].EndsWith("\x03"))
                    fields[14] = fields[14].Substring(0, fields[14].Length - 1);
            }
            else
            {
                // Probably really long article title or something that got cut off; we can't handle these
                return;
            }

            try
            {
                RCEvent rce;
                rce.eventtype = RCEvent.EventType.unknown;
                rce.blockLength = "";
                rce.movedTo = "";
                rce.project = e.Data.Channel.Substring(1);
                rce.title = Project.TranslateNamespace(rce.project, fields[2]);
                rce.url = fields[6];
                rce.user = fields[10];
                Project project = ((Project)Program.prjlist[rce.project]);
                // At the moment, fields[14] contains IRC colour codes. For plain edits, remove just the \x03's. For logs, remove using the regex.
                Match titlemo = project.rSpecialLogRegex.Match(fields[2]);
                if (!titlemo.Success)
                {
                    // This is a regular edit
                    rce.minor = fields[4].Contains("M");
                    rce.newpage = fields[4].Contains("N");
                    rce.botflag = fields[4].Contains("B");
                    rce.eventtype = RCEvent.EventType.edit;
                    rce.comment = fields[14].Replace("\x03", "");
                }
                else
                {
                    // This is a log edit; check for type
                    string logType = titlemo.Groups[1].Captures[0].Value;
                    // Fix comments
                    rce.comment = stripColours2.Replace(fields[14], "");
                    switch (logType)
                    {
                        case "newusers":
                            // Could be a user creating their own account, or a user creating a sockpuppet

                            // Example message as of 2016-11-02 on #nl.wikipedia (with log comment after colon)
                            // > [[Speciaal:Log/newusers]] create2  * BRPots *  created new account Gebruiker:BRPwiki: eerder fout gemaakt
                            // Example message as of 2016-11-02 on #nl.wikipedia (without log comment)
                            // > [[Speciaal:Log/newusers]] create2  * Sherani koster *  created new account Gebruiker:Rani farah koster

                            // Example message as of 2017-10-13 on #en.wikipedia:
                            // > [[Special:Log/newusers]] create2  * Ujju.19788 *  created new account User:Upendhare
                            // Example message as of 2022-01-24 on #en.wikipedia:
                            // > [[Special:Log/newusers]] byemail  * Mdaniels5757 *  created new account User:Hannahco12: Requested account
                            //
                            // Treat newusers/byemail the same as newusers/create2.
                            // MediaWiki internally re-uses the "create2" message for "byemail" as well.
                            // Ref mediawiki-core.git:/LogFormatter.php#getIRCActionText
                            // Ref https://phabricator.wikimedia.org/T327126
                            //
                            if (fields[4].Contains("create2") || fields[4].Contains("byemail"))
                            {
                                Match mc2 = project.rCreate2Regex.Match(rce.comment);
                                if (mc2.Success)
                                {
                                    rce.title = mc2.Groups[1].Captures[0].Value;
                                    rce.eventtype = RCEvent.EventType.newuser2;
                                }
                                else
                                {
                                    logger.Warn("Unmatched create2 event in " + rce.project + ": " + e.Data.Message);
                                }
                            }
                            else
                            {
                                if (fields[4].Contains("autocreate"))
                                {
                                    rce.eventtype = RCEvent.EventType.autocreate;
                                }
                                else
                                {
                                    rce.eventtype = RCEvent.EventType.newuser;
                                }
                            }
                            break;
                        case "block":
                            // Example message from October 2017 on #en.wikipedia:
                            // > [[Special:Log/block]] reblock  * Yamla *  changed block settings for [[User:Jeb BushDid911]] (account creation blocked, email disabled, cannot edit own talk page) with an expiry time of indefinite: {{uw-ublock}}
                            //
                            // Example message from October 2017 on #en.wikipedia
                            // > [[Special:Log/block]] reblock  * DeltaQuad *  changed block settings for [[User:208.111.64.0/19]] (anon. only, account creation blocked) with an expiry time of 06:21, February 2, 2019: {{colocationwebhost}}
                            if (fields[4].Contains("unblock"))
                            {
                                Match ubm = project.runblockRegex.Match(rce.comment);
                                if (ubm.Success)
                                {
                                    rce.eventtype = RCEvent.EventType.unblock;
                                    rce.title = ubm.Groups["item1"].Captures[0].Value;
                                    try
                                    {
                                        rce.comment = ubm.Groups["comment"].Captures[0].Value;
                                    }
                                    catch (ArgumentOutOfRangeException) { }
                                }
                                else
                                {
                                    logger.Warn("Unmatched block/unblock type in " + rce.project + ": " + e.Data.Message);
                                    return;
                                }
                            }
                            else if (fields[4].Contains("reblock"))
                            {
                                Match rbm = project.rreblockRegex.Match(rce.comment);
                                if (rbm.Success)
                                {
                                    // Treat reblock the same as a new block for simplicity
                                    rce.eventtype = RCEvent.EventType.block;
                                    rce.title = rbm.Groups["item1"].Captures[0].Value;
                                }
                                else
                                {
                                    logger.Warn("Unmatched block/reblock type in " + rce.project + ": " + e.Data.Message);
                                    return;
                                }
                            }
                            else
                            {
                                Match bm = project.rblockRegex.Match(rce.comment);
                                if (bm.Success)
                                {
                                    rce.eventtype = RCEvent.EventType.block;
                                    rce.title = bm.Groups["item1"].Captures[0].Value;
                                    // Assume default value of 24 hours in case the on-wiki message override
                                    // is missing expiry ($2) from its interface messag
                                    rce.blockLength = "24 hours";
                                    try
                                    {
                                        rce.blockLength = bm.Groups["item2"].Captures[0].Value;
                                    }
                                    catch (ArgumentOutOfRangeException) { }
                                    try
                                    {
                                        rce.comment = bm.Groups["comment"].Captures[0].Value;
                                    }
                                    catch (ArgumentOutOfRangeException) { }
                                }
                                else
                                {
                                    logger.Warn("Unmatched block type in " + rce.project + ": " + e.Data.Message);
                                    return;
                                }
                            }
                            break;
                        case "protect":
                            // Could be a protect, modifyprotect or unprotect; need to parse regex
                            Match pm = project.rprotectRegex.Match(rce.comment);
                            Match modpm = project.rmodifyprotectRegex.Match(rce.comment);
                            Match upm = project.runprotectRegex.Match(rce.comment);
                            if (pm.Success)
                            {
                                rce.eventtype = RCEvent.EventType.protect;
                                rce.title = Project.TranslateNamespace(rce.project, pm.Groups["item1"].Captures[0].Value);
                                try
                                {
                                    rce.comment = pm.Groups["comment"].Captures[0].Value;
                                }
                                catch (ArgumentOutOfRangeException) { }
                            }
                            else if (modpm.Success)
                            {
                                rce.eventtype = RCEvent.EventType.modifyprotect;
                                rce.title = Project.TranslateNamespace(rce.project, modpm.Groups["item1"].Captures[0].Value);
                                try
                                {
                                    rce.comment = modpm.Groups["comment"].Captures[0].Value;
                                }
                                catch (ArgumentOutOfRangeException) { }
                            }
                            else
                            {
                                if (upm.Success)
                                {
                                    rce.eventtype = RCEvent.EventType.unprotect;
                                    rce.title = Project.TranslateNamespace(rce.project, upm.Groups["item1"].Captures[0].Value);
                                    try
                                    {
                                        rce.comment = upm.Groups["comment"].Captures[0].Value;
                                    }
                                    catch (ArgumentOutOfRangeException) { }
                                }
                                else
                                {
                                    logger.Warn("Unmatched protect type in " + rce.project + ": " + e.Data.Message);
                                    return;
                                }
                            }
                            break;
                        case "rights":
                            // Ignore event
                            return;
                        //break;
                        case "delete":
                            // Could be a delete or restore; need to parse regex
                            Match dm = project.rdeleteRegex.Match(rce.comment);
                            if (dm.Success)
                            {
                                rce.eventtype = RCEvent.EventType.delete;
                                rce.title = Project.TranslateNamespace(rce.project, dm.Groups["item1"].Captures[0].Value);
                                try
                                {
                                    rce.comment = dm.Groups["comment"].Captures[0].Value;
                                }
                                catch (ArgumentOutOfRangeException) { }
                            }
                            else
                            {
                                Match udm = project.rrestoreRegex.Match(rce.comment);
                                if (udm.Success)
                                {
                                    rce.eventtype = RCEvent.EventType.restore;
                                    rce.title = Project.TranslateNamespace(rce.project, udm.Groups["item1"].Captures[0].Value);
                                    try
                                    {
                                        rce.comment = udm.Groups["comment"].Captures[0].Value;
                                    }
                                    catch (ArgumentOutOfRangeException) { }
                                }
                                else
                                {
                                    // Could be 'revision' (change visibility of revision) or something else
                                    // Ignore event
                                    return;
                                }
                            }
                            break;
                        case "upload":
                            Match um = project.ruploadRegex.Match(rce.comment);
                            if (um.Success)
                            {
                                rce.eventtype = RCEvent.EventType.upload;
                                rce.title = Project.TranslateNamespace(rce.project, um.Groups["item1"].Captures[0].Value);
                                try
                                {
                                    rce.comment = um.Groups["comment"].Captures[0].Value;
                                }
                                catch (ArgumentOutOfRangeException) { }
                            }
                            else
                            {
                                // Could be 'overwrite' (upload new version) or something else
                                // Ignore event
                                return;
                            }
                            break;
                        case "move":
                            //Is a move
                            rce.eventtype = RCEvent.EventType.move;
                            //Check "move over redirect" first: it's longer, and plain "move" may match both (e.g., en-default)
                            Match mrm = project.rmoveredirRegex.Match(rce.comment);
                            if (mrm.Success)
                            {
                                rce.title = Project.TranslateNamespace(rce.project, mrm.Groups["item1"].Captures[0].Value);
                                rce.movedTo = Project.TranslateNamespace(rce.project, mrm.Groups["item2"].Captures[0].Value);
                                //We use the unused blockLength field to store our "moved from" URL
                                rce.blockLength = project.rooturl + "wiki/" + CVNBotUtils.WikiEncode(mrm.Groups["item1"].Captures[0].Value);
                                try
                                {
                                    rce.comment = mrm.Groups["comment"].Captures[0].Value;
                                }
                                catch (ArgumentOutOfRangeException) { }
                            }
                            else
                            {
                                Match mm = project.rmoveRegex.Match(rce.comment);
                                if (mm.Success)
                                {
                                    rce.title = Project.TranslateNamespace(rce.project, mm.Groups["item1"].Captures[0].Value);
                                    rce.movedTo = Project.TranslateNamespace(rce.project, mm.Groups["item2"].Captures[0].Value);
                                    //We use the unused blockLength field to store our "moved from" URL
                                    rce.blockLength = project.rooturl + "wiki/" + CVNBotUtils.WikiEncode(mm.Groups["item1"].Captures[0].Value);
                                    try
                                    {
                                        rce.comment = mm.Groups["comment"].Captures[0].Value;
                                    }
                                    catch (ArgumentOutOfRangeException) { }
                                }
                                else
                                {
                                    logger.Warn("Unmatched move type in " + rce.project + ": " + e.Data.Message);
                                    return;
                                }
                            }
                            break;
                        case "import":
                            //rce.eventtype = RCEvent.EventType.import;
                            // Ignore event
                            return;
                        //break;
                        default:
                            // Ignore event
                            return;
                    }
                    // These flags don't apply to log events, but must be initialized
                    rce.minor = false;
                    rce.newpage = false;
                    rce.botflag = false;
                }

                // Deal with the diff size
                Match n = rszDiff.Match(fields[13]);
                if (n.Success)
                {
                    if (n.Groups[1].Captures[0].Value == "+")
                        rce.szdiff = Convert.ToInt32(n.Groups[2].Captures[0].Value);
                    else
                        rce.szdiff = 0 - Convert.ToInt32(n.Groups[2].Captures[0].Value);
                }
                else
                    rce.szdiff = 0;

                try
                {
                    Program.ReactToRCEvent(rce);
                }
                catch (Exception exce)
                {
                    logger.Error("Failed to handle RCEvent", exce);
                    Program.BroadcastDD("ERROR", "ReactorException", exce.Message, e.Data.Channel + " " + e.Data.Message);
                }
            }
            catch (ArgumentOutOfRangeException eor)
            {
                // Broadcast this for Distributed Debugging
                logger.Error("Failed to process incoming message", eor);
                Program.BroadcastDD("ERROR", "RCR_AOORE", eor.Message, e.Data.Channel + "/" + e.Data.Message
                    + "Fields: " + fields);
            }
        }

    }
}

File: cvnbot/rcreader.py
import dataclasses
import datetime
import enum
import logging
import re
import threading

from . import utils
from .ircclient import IrcClient, IrcConnectionError

logger = logging.getLogger("CVNBot.RCReader")


class EventType(enum.Enum):
    delete = "delete"
    restore = "restore"
    upload = "upload"
    block = "block"
    unblock = "unblock"
    edit = "edit"
    protect = "protect"
    unprotect = "unprotect"
    move = "move"
    rollback = "rollback"
    newuser = "newuser"
    imported = "import"
    unknown = "unknown"
    newuser2 = "newuser2"
    autocreate = "autocreate"
    modifyprotect = "modifyprotect"


@dataclasses.dataclass
class RCEvent:
    project: str = ""
    title: str = ""
    url: str = ""
    user: str = ""
    minor: bool = False
    newpage: bool = False
    botflag: bool = False
    szdiff: int = 0
    comment: str = ""
    eventtype: EventType = EventType.unknown
    block_length: str = ""
    moved_to: str = ""


class RCReader:
    """Connect to the Wikimedia RC feed and forward events to Program."""

    # RC parsing regexen
    strip_colours = re.compile(r"\x04\d{0,2}\*?")
    strip_colours2 = re.compile(r"\x03\d{0,2}")
    strip_bold = re.compile(r"\x02")
    rsz_diff = re.compile(r"\(([\+\-])([0-9]+)\)")

    SERVER_NAME = "irc.wikimedia.org"

    def __init__(self, program):
        self.bot = program
        self.rcirc = IrcClient("RCReader",
                               auto_reconnect=True,
                               auto_rejoin=True,
                               )
        self.last_message = datetime.datetime.now()

    def initiate_connection(self):
        """Connect, join every monitored wiki's channel, and read forever."""
        threading.current_thread().name = "RCReader"

        logger.info("Thread started")

        # Set up RCReader
        self.rcirc.on_connected = self._on_connected
        self.rcirc.on_channel_message = self._on_channel_message

        try:
            self.rcirc.connect(self.SERVER_NAME, 6667)
        except IrcConnectionError as e:
            logger.warning("Could not connect: %s", e)
            return

        self.rcirc.login(self.bot.config.bot_nick, "CVNBot", 4, "CVNBot")

        logger.info("Joining %d channels", len(self.bot.prjlist))
        for project_name in self.bot.prjlist.keys():
            self.rcirc.rfc_join("#" + project_name)

        # Enter loop
        self.rcirc.listen()
        # When listen() returns the IRC session is over
        self.rcirc.disconnect()

    def _on_connected(self, client):
        logger.info("Connected to %s", self.SERVER_NAME)

    def _on_channel_message(self, client, event):
        self.last_message = datetime.datetime.now()

        rce = self.parse_message(event.channel, event.message)
        if rce is None:
            return

        try:
            self.bot.react_to_rc_event(rce)
        except Exception as e:
            logger.exception("Failed to handle RCEvent")
            self.bot.broadcast_dd(
                "ERROR", "ReactorException", str(e),
                event.channel + " " + event.message,
            )

    def parse_message(self, channel, message):
        """
        Parse one raw line into an RCEvent, or None to ignore it.

        Sample from #en.wikipedia on 2017-10-13:
          01> #00314 [[
          02> #00307 Special:Log/newusers
          03> #00314 ]]
          04> #0034   create2
          05> #00310
          06> #00302
          07> #003
          08> #0035  *
          09> #003
          10> #00303 Ujju.19788
          11> #003
          12> #0035  *
          13> #003
          14> #00310 created new account User:Upendhare
          15> #003
        """
        stripped = message
        stripped = utils.replace_str_max(stripped, "\x03", "\x04", 14)
        stripped = self.strip_colours.sub("\x03", stripped)
        stripped = self.strip_bold.sub("", stripped)
        fields = stripped.split("\x03", 14)
        if len(fields) == 15:
            if fields[14].endswith("\x03"):
                fields[14] = fields[14][:-1]
        else:
            # Probably really long article title or something that got cut off;
            # we can't handle these
            return None

        # "#en.wikipedia" > "en.wikipedia"
        project_name = channel[1:]
        if project_name not in self.bot.prjlist:
            logger.warning("Failed to process incoming from %s\n%s\n%s", channel, message, fields)
            self.bot.broadcast_dd(
                "ERROR", "RCR_AOORE", "Message for unmonitored project",
                "{0} {1}".format(channel, stripped),
            )
            # Ignore
            return None
        project = self.bot.prjlist[project_name]

        rce = RCEvent(
            project=project_name,
            title=project.translate_namespace(fields[2]),
            url=fields[6],
            user=fields[10],
        )

        # At the moment, fields[14] contains IRC colour codes.
        # For plain edits, remove just the \x03's.
        # For logs, remove using the regex.
        titlemo = project.rspecial_log_regex.search(fields[2])
        if not titlemo:
            # This is a regular edit
            rce.minor = "M" in fields[4]
            rce.newpage = "N" in fields[4]
            rce.botflag = "B" in fields[4]
            rce.eventtype = EventType.edit
            rce.comment = fields[14].replace("\x03", "")
        else:
            # This is a log action
            log_type = titlemo.group(1)
            # Fix comments
            rce.comment = self.strip_colours2.sub("", fields[14])
            # Check logevent type
            if not self._parse_log_event(rce, project, log_type, fields, message):
                return None

            # These flags don't apply to log events, but must be initialized
            rce.minor = False
            rce.newpage = False
            rce.botflag = False

        # Deal with the diff size
        diff = self.rsz_diff.search(fields[13])
        if diff:
            rce.szdiff = int(diff.group(2))
            if diff.group(1) == "-":
                rce.szdiff = -rce.szdiff
        else:
            rce.szdiff = 0

        return rce

    def _parse_log_event(self, rce, project, log_type, fields, message):
        """Set rce.eventtype and possibly other fields; returns False when the event must be ignored."""
        if log_type == "newusers":
            return self._parse_newusers(rce, project, fields, message)
        if log_type == "block":
            return self._parse_block(rce, project, fields, message)
        if log_type == "protect":
            return self._parse_protect(rce, project, message)
        if log_type == "delete":
            return self._parse_delete(rce, project)
        if log_type == "upload":
            return self._parse_upload(rce, project)
        if log_type == "move":
            return self._parse_move(rce, project, message)

        # Ignore "rights", "import", and any other log event
        return False

    # Could be a user creating their own account, or a user creating a sockpuppet
    #
    # Sample from #nl.wikipedia in 2016 (with comment)
    #   [[Speciaal:Log/newusers]] create2  * BRPots *  created new account Gebruiker:BRPwiki: eerder fout gemaakt
    #
    # Sample from #nl.wikipedia in 2016 (without comment)
    #   [[Speciaal:Log/newusers]] create2  * Sherani koster *  created new account Gebruiker:Rani farah koster
    #
    # Sample from #en.wikipedia in 2017:
    #   [[Special:Log/newusers]] create2  * Ujju.19788 *  created new account User:Upendhare
    #
    # Sample from #en.wikipedia in 2022:
    #   [[Special:Log/newusers]] byemail  * Mdaniels5757 *  created new ccount User:Hannahco12: Requested account
    #
    # Treat newusers/byemail the same as newusers/create2.
    # MediaWiki internally re-uses the "create2" message for "byemail" as well.
    # Ref mediawiki-core.git:/LogFormatter.php#getIRCActionText
    # Ref https://phabricator.wikimedia.org/T327126
    #
    # See also Program._react_newuser2, which formats creator=user, and editor=title
    def _parse_newusers(self, rce, project, fields, message):
        if "create2" in fields[4] or "byemail" in fields[4]:
            match = project.rcreate2_regex.search(rce.comment)
            if not match:
                logger.warning("Unmatched create2 event in %s: %s", rce.project, message)
                return False
            rce.title = match.group(1)
            rce.eventtype = EventType.newuser2
            return True

        if "autocreate" in fields[4]:
            rce.eventtype = EventType.autocreate
        else:
            rce.eventtype = EventType.newuser
        return True

    # Sample from #test.wikipedia in August 2026:
    # [[Special:Log/block]] block  * Krinkle *  blocked [[User:KrinkleBot]] with an expiration time of 5 minutes (autoblock disabled)
    # [[Special:Log/block]] block  * Krinkle *  blocked [[User:KrinkleBot]] with an expiration time of 00:00, 25 August 2026 (account creation disabled)
    # [[Special:Log/block]] reblock  * Krinkle *  changed block settings for [[User:KrinkleBot]] with an expiration time of 1 second (autoblock disabled)
    # [[Special:Log/block]] reblock  * Krinkle *  changed block settings for [[User:KrinkleBot]] with an expiration time of 1 second (autoblock disabled): example {{here}}
    # [[Special:Log/block]] reblock  * Krinkle *  changed block settings for [[User:KrinkleBot]] with an expiration time of 20:00, 24 August 2026 (autoblock disabled, email disabled, cannot edit own talk page)
    # [[Special:Log/block]] reblock  * Krinkle *  changed block settings for [[User:KrinkleBot]] with an expiration time of 20:00, 24 August 2026 (autoblock disabled, email disabled, cannot edit own talk page): example
    # [[Special:Log/block]] unblock  * Krinkle *  unblocked User:KrinkleBot
    def _parse_block(self, rce, project, fields, message):
        if "unblock" in fields[4]:
            match = project.runblock_regex.search(rce.comment)
            if not match:
                logger.warning("Unmatched block/unblock in %s: %s", rce.project, message)
                return False
            rce.eventtype = EventType.unblock
            rce.title = match.group("item1")
            # This project regex might not have a "comment" match group
            rce.comment = match.groupdict(default="").get("comment", "")
            return True

        if "reblock" in fields[4]:
            match = project.rreblock_regex.search(rce.comment)
            if not match:
                logger.warning("Unmatched block/reblock in %s: %s", rce.project, message)
                return False
            # Treat reblock the same as a new block for simplicity
            rce.eventtype = EventType.block
            rce.title = match.group("item1")
            return True

        match = project.rblock_regex.search(rce.comment)
        if not match:
            logger.warning("Unmatched block type in %s: %s", rce.project, message)
            return False

        rce.eventtype = EventType.block
        rce.title = match.group("item1")
        # Assume default value of 24 hours in case the on-wiki message override
        # is missing expiry ($2) from its interface message
        item2 = match.groupdict().get("item2")
        rce.block_length = item2 if item2 is not None else "24 hours"
        # This project regex might not have a "comment" match group
        rce.comment = match.groupdict(default="").get("comment", "")
        return True

    # Sample from #en.wikipedia in August 2026:
    # [[Special:Log/protect]] protect  * Codename Noreste *  protected "[[Module:Message box/tmbox.css ‎[edit=sysop] (indefinite)‎[move=sysop] (indefinite)]]": Highly visible template
    #
    # Sample from #he.wikipedia in August 2026:
    # [[מיוחד:Log/protect]] protect  * Krinkle *  protected "[[ויקיפדיה:ארגז חול ‏[edit=autoconfirmed] (פגה ב־05:43, 25 באוגוסט 2026 (UTC))‏[move=autoconfirmed] (פגה ב־05:43, 25 באוגוסט 2026 (UTC))]]"
    def _parse_protect(self, rce, project, message):
        # TODO: Fix rce.title for protect and modifyprotect events
        # This wrongly contains "[move=sysop] (indefinite)" etc.
        # In mediawiki-core/LogFormatter.php#getIRCActionText it uses 'protectedarticle'
        # ('protected "[[$1]]"') with `$target . ' ' . $parameters['4::description']` as parameter.
        #
        # The good news is that mediawiki-core/ProtectLogFormatter.php prepends
        # the restrictions addendum with Language::getDirMark, i.e. 0x200e (U+200E LEFT-TO-RIGHT MARK)
        # or 0x200f (U+200F RIGHT-TO-LEFT MARK).
        #
        # See test_rcreader.py#test_protect_real_ltr_comment
        for regex, eventtype in (
            (project.rprotect_regex, EventType.protect),
            (project.rmodifyprotect_regex, EventType.modifyprotect),
            (project.runprotect_regex, EventType.unprotect),
        ):
            match = regex.search(rce.comment)
            if match:
                rce.eventtype = eventtype
                rce.title = project.translate_namespace(match.group("item1"))
                # This project regex might not have a "comment" match group
                rce.comment = match.groupdict(default="").get("comment", "")
                return True

        logger.warning("Unmatched protect type in %s: %s", rce.project, message)
        return False

    def _parse_delete(self, rce, project):
        for regex, eventtype in (
            (project.rdelete_regex, EventType.delete),
            (project.rrestore_regex, EventType.restore),
        ):
            match = regex.search(rce.comment)
            if match:
                rce.eventtype = eventtype
                rce.title = project.translate_namespace(match.group("item1"))
                # This project regex might not have a "comment" match group
                rce.comment = match.groupdict(default="").get("comment", "")
                return True

        # Ignore "delete/revision" (change visibility of revision) and any other deletion events
        return False

    def _parse_upload(self, rce, project):
        match = project.rupload_regex.search(rce.comment)
        if not match:
            # Ignore "upload/overwrite" (upload new version) and any other upload event
            return False
        rce.eventtype = EventType.upload
        rce.title = project.translate_namespace(match.group("item1"))
        # This project regex might not have a "comment" match group
        rce.comment = match.groupdict(default="").get("comment", "")
        return True

    def _parse_move(self, rce, project, message):
        rce.eventtype = EventType.move
        # First check "move over redirect", because it uses a longer format,
        # whereas plain "move" may match both.
        for regex in (project.rmoveredir_regex, project.rmove_regex):
            match = regex.search(rce.comment)
            if match:
                rce.title = project.translate_namespace(match.group("item1"))
                rce.moved_to = project.translate_namespace(match.group("item2"))
                # We use the unused block_length field to store our "moved from" URL
                rce.block_length = project.rooturl + "wiki/" + utils.wiki_encode(match.group("item1"))
                # This project regex might not have a "comment" match group
                rce.comment = match.groupdict(default="").get("comment", "")
                return True

        logger.warning("Unmatched move type in %s: %s", rce.project, message)
        return False
